src/main/java/hudson/remoting/Engine.java

```diff
@@ -239,40 +232,36 @@ public void run() {
 
                 events.status("Handshaking");
                 Socket jnlpSocket = connect(host,port);
-                DataOutputStream outputStream = new DataOutputStream(jnlpSocket.getOutputStream());
-                BufferedInputStream inputStream = new BufferedInputStream(jnlpSocket.getInputStream());
-                boolean connected = false;
+                ChannelBuilder channelBuilder = new ChannelBuilder("channel", executor)
+                        .withJarCache(jarCache)
+                        .withMode(Mode.BINARY);
+                Channel channel = null;
 
                 // Try available protocols.
                 for (JnlpProtocol protocol : protocols) {
                     events.status("Trying protocol: " + protocol.getName());
-                    String response = protocol.performHandshake(outputStream, inputStream);
+                    try {
+                        channel = protocol.establishChannel(jnlpSocket, channelBuilder);
+                    } catch (IOException ioe) {
+                        events.status("Protocol failed to establish channel", ioe);
+                    }
 
                     // On success do not try other protocols.
-                    if (response.equals(GREETING_SUCCESS)) {
-                        connected = true;
+                    if (channel != null) {
                         break;
                     }
 
-                    // On failure log the response and form a new connection.
-                    events.status("Server didn't understand the protocol: " + response);
+                    // On failure form a new connection.
                     jnlpSocket.close();
                     jnlpSocket = connect(host,port);
-                    outputStream = new DataOutputStream(jnlpSocket.getOutputStream());
-                    inputStream = new BufferedInputStream(jnlpSocket.getInputStream());
                 }
 
                 // If no protocol worked.
-                if (!connected) {
+                if (channel == null) {
                     onConnectionRejected("None of the protocols were accepted");
                     continue;
                 }
 
-                final Channel channel = new ChannelBuilder("channel", executor)
-                        .withJarCache(jarCache)
-                        .withMode(Mode.BINARY)
-                        .build(inputStream, new BufferedOutputStream(jnlpSocket.getOutputStream()));
```

Comment:
@oleg-nenashev the old code would keep the BufferedInputStream to be used when creating the channel, with this change however, the buffered input stream (and it's read-ahead buffer) is thrown away because [establishChannel](https://github.com/jenkinsci/remoting/pull/41/files#diff-2743d6a321ae5abd61a9ec59c4488a95R72) throws the Buffer away